### ISSUE TITLE
Sharded Replication fails due to no node data.

### ISSUE THREAD

**Issue by dgarstang** [edelight/chef-mongodb#298]
I'm trying to set up two nodes in the same shard with replication. 

When the chef client is run on either of the sharded nodes, it fails. The recipe references data on the other node. However, since the chef-client has not run to completion on the other node yet, the keys attempted to be accessed in the node data don't exist yet. Catch-22. Deadlock. Obviously it doesn't matter which node you run chef-client on. Both will fail.

I'm not a ruby expert, but some debug statements added to library/mongodb.rb show that it's getting a list of members (i.e. myself and the other member of the shard) and then trying to iterate over those and pull node data. 

This seems like a ... rather large problem. I know its a limitation of how chef works, but ...

If there's a workaround of some sort... would love to know...

**Comment by jamesonjlee** [comment]
have you tried setting up shards only then adding the replica sets?

**Comment by dgarstang** [comment]
No. Do you think that might be a workaround? Obviously, would need to provision the sharded servers twice... once without sharding, then modifying the node on the chef server and/or the runlist and running the chef client again to reprovision.

**Comment by srathbun** [comment]
I bumped into this too. It's especially annoying because [this line](https://github.com/edelight/chef-mongodb/blob/master/libraries/mongodb.rb#L62) fails since there is no slave_delay on the node, and the previous lookup gets `nil:NilClass`. The same thing happens on [this line](https://github.com/edelight/chef-mongodb/blob/master/libraries/mongodb.rb#L69) as well.

That kills the chef-client update, so the nodes never get their defaults. If these are set specifically on the nodes, then it fails to initiate the replica set, since it can't get the default port, but it does not fail the update. Since the update didn't fail, running it again works successfully, because it has picked up the defaults it needs, like the mongo port.

Perhaps a bootstrap recipe that could be run first to configure defaults on the nodes?

**Comment by jamesonjlee** [comment]
that was the intent of the default recipe that drop in the configurations, yeah the empty members for setting up replica/shards is still an ongoing issue.

what should the bootstrap recipe do that the current default install recipes doesn't do?

**Comment by srathbun** [comment]
Basically, instead of bringing up mongo or installing a service, it would just set properties on the node. A user would not add it to the run list for this node, but would override the a chef client run to set all these defaults. I.E. `knife ssh 'role:shard' 'chef-client -o bootstrap'`

That would get them the attributes they need for when the client actually runs and brings mongo up. The trouble is that without a successful chef run, the node will not be updated for chef search to find. Thus, it needs to be done manually and a bootstrap script is a nicety. Otherwise the user has to go to each node and configure a bunch of options themselves, or run twice understanding that it will fail the first time.

**Comment by jamesonjlee** [comment]
@srathbun ah yes the 'if 3 instances come up at the same time' problem.  The best solution for that in my opinion is to not to do it inside of chef (Chef isn't responsible for orchestration).  There are potential solutions (like doing a separate discovery channel) but they need additional work to setup.

My usual recommendation for this situation is to spin up 1 node first, let it converge then begin spinning up additional nodes (one at a time).  One alternative (which someone can open a PR for) is to code in a alternative for member nodes via an attribute (so you can hard code in your node members)
